### PR TITLE
feat: Add a flag to the `uv_spawn` to trigger UAC on Windows

### LINKED ISSUE #4295: support launching executables with runas permissions
Calling UAC applications requires a complicated, undocumented RPC call, or using the `runas` verb to ShellExecuteEx to do that for the program. Perhaps this should be a flag to `uv_spawn`, so that users can run processes such as `regedit.exe` and `setup.exe` (after clicking on the UAC prompt) with libuv, without requiring workarounds such as `cmd /c regedit`, which will run the ShellExecute function verb "open"? See https://learn.microsoft.com/en-us/windows/win32/shell/launch

### PR DESCRIPTION
**Pull request by CGQAQ**
This PR handles
*  `UV_PROCESS_WINDOWS_HIDE_CONSOLE` and `UV_PROCESS_WINDOWS_HIDE`
*  `UV_PROCESS_WINDOWS_HIDE_GUI` and `UV_PROCESS_WINDOWS_HIDE`

other flags ARE ignored silently due to `ShellExecuteExW` limitations

REFS
* [SHELLEXECUTEINFOW](https://learn.microsoft.com/en-us/windows/win32/api/shellapi/ns-shellapi-shellexecuteinfow)
* [ShellExecuteExW](https://learn.microsoft.com/en-us/windows/win32/api/shellapi/nf-shellapi-shellexecuteexw)
* [Launching Applications](https://learn.microsoft.com/en-us/windows/win32/shell/launch)

closes: #4295

### CODE REVIEW

**Comment by CGQAQ** [comment]
@vtjnash done! PTAL

**Comment by vtjnash** [comment]
Seems good to me, what do other libuv devs think? It is platform specific, but so are a lot of the existing flags here. It will need a test and addition to the docs. And some of the options should probably error if passed in combination with it, if the combination is not feasible.

**Comment by santigimeno** [comment]
> what do other libuv devs think?

SGTM

**Comment by CGQAQ** [comment]
> It will need a test 
@vtjnash any ideas on how to test this? once you trigger the UAC on subprocess, you need user's interaction to either click allow or disallow

File: src/win/process.c [hunk 1]
```diff
@@ -1069,27 +1067,62 @@ int uv_spawn(uv_loop_t* loop,
...
+    shellexecuteinfo.fMask = fMask | SEE_MASK_WAITFORINPUTIDLE  | SEE_MASK_NOCLOSEPROCESS; // for shellexecuteinfo.hProcess
+    shellexecuteinfo.lpVerb = L"runas";
+    shellexecuteinfo.lpFile = application_path;
+    shellexecuteinfo.lpParameters = arguments;
+    shellexecuteinfo.nShow = startup.wShowWindow;
+    if (!ShellExecuteExW(&shellexecuteinfo)) {
+        err = GetLastError();
+        goto done;
+    }
+    hProc = shellexecuteinfo.hProcess;
```
Comment (vtjnash) [inline src/win/process.c:1081]: The documentation states this sometimes returns NULL instead of the hProcess. How should libuv deal with that?

Comment (CGQAQ) [reply src/win/process.c:1081]: > Even if fMask is set to SEE_MASK_NOCLOSEPROCESS, hProcess will be NULL if no process was launched

Is no process actually have been launched considered an error?

File: src/win/process.c [hunk 2]
```diff
@@ -1073,27 +1071,62 @@ int uv_spawn(uv_loop_t* loop,
...
-                     process_flags,
-                     env,
-                     cwd,
-                     &startup,
-                     &info)) {
-    /* CreateProcessW failed. */
-    err = GetLastError();
-    goto done;
+  if (options->flags & UV_PROCESS_WINDOWS_RUNAS_ADMIN) {
+    shellexecuteinfo.cbSize = sizeof(SHELLEXECUTEINFOW);
```
Comment (bnoordhuis) [inline src/win/process.c:1075]:
```suggestion
    shellexecuteinfo.cbSize = sizeof(shellexecuteinfo);
```

File: src/win/process.c (hunk 1 above)
Comment (bnoordhuis) [reply src/win/process.c:1085]: On Unix systems you'd get an error like ENOENT.

File: src/win/process.c [hunk 3]
```diff
@@ -1073,27 +1071,62 @@ int uv_spawn(uv_loop_t* loop,
...
+    shellexecuteinfo.lpVerb = L"runas";
+    shellexecuteinfo.lpFile = application_path;
+    shellexecuteinfo.lpParameters = arguments;
+    shellexecuteinfo.nShow = startup.wShowWindow;
+    if (!ShellExecuteExW(&shellexecuteinfo)) {
+        err = GetLastError();
+        goto done;
+    }
+    hProc = shellexecuteinfo.hProcess;
+    if ((dwProcId = GetProcessId(hProc)) == 0) {
```
Comment (bnoordhuis) [inline src/win/process.c:1086]:
```suggestion
    dwProcId = GetProcessId(hProc);
    if (dwProcId == 0) {
```

File: src/win/process.c [hunk 4]
```diff
@@ -1073,27 +1071,62 @@ int uv_spawn(uv_loop_t* loop,
...
+        /* CreateProcessW failed. */
+        err = GetLastError();
+        goto done;
+    }
+    hProc = info.hProcess;
+    dwProcId = info.dwProcessId;
+    hThread = info.hThread;
   }
 
+
```
Comment (bnoordhuis) [inline src/win/process.c:1123]:
```suggestion
```

File: src/win/process.c [hunk 5]
```diff
@@ -1110,18 +1143,18 @@ int uv_spawn(uv_loop_t* loop,
     }
   }
 
-  if (process_flags & CREATE_SUSPENDED) {
-    if (ResumeThread(info.hThread) == ((DWORD)-1)) {
+  if (process_flags & CREATE_SUSPENDED && hThread != NULL) {
```
Comment (bnoordhuis) [inline src/win/process.c:1146]:
```suggestion
  if ((process_flags & CREATE_SUSPENDED) && hThread != NULL) {
```
Having said that.... UV_PROCESS_DETACHED and UV_PROCESS_WINDOWS_RUNAS_ADMIN should probably be mutually exclusive if they don't work together?

File: src/win/process.c [hunk 6]
```diff
@@ -917,6 +918,13 @@ int uv_spawn(uv_loop_t* loop,
          *env = NULL, *cwd = NULL;
   STARTUPINFOW startup;
   PROCESS_INFORMATION info;
+  SHELLEXECUTEINFOW shellexecuteinfo = {0};
+
+  HANDLE hProc;
+  DWORD  dwProcId;
```
[the remark this answers is not in the document]
Comment (CGQAQ) [reply src/win/process.c:924]: I see

File: src/win/process.c (hunk 1 above)
Comment (CGQAQ) [reply src/win/process.c]: done

File: src/win/process.c (hunk 5 above)
Comment (CGQAQ) [reply src/win/process.c]: how? is that means if two specified together, we return an error OR we ignore one of them, if the former one, which error should be returned?

Comment (bnoordhuis) [reply src/win/process.c]: Specifying them both should be an error (and should be documented as such), probably UV_EINVAL.

Comment (CGQAQ) [reply src/win/process.c]: done

File: src/win/process.c [hunk 7]
```diff
@@ -1073,27 +1077,66 @@ int uv_spawn(uv_loop_t* loop,
...
+                        arguments,
+                        NULL,
+                        NULL,
+                        1,
+                        process_flags,
+                        env,
+                        cwd,
+                        &startup,
+                        &info)) {
+        /* CreateProcessW failed. */
```
Comment (bnoordhuis) [inline src/win/process.c:1125]:
```suggestion
```
Superfluous

**Review by bnoordhuis — DISMISSED** [review]
Thanks, LGTM at a quick glance. @vtjnash want to look this over one more time?

**Review by bnoordhuis — COMMENTED** [review]
I'm taking back my LGTM for the moment because it silently ignores stdio redirection and other options, doesn't it? Incompatible options should be errors instead.

**Comment by CGQAQ** [comment]
@bnoordhuis resolved, PTAL

File: docs/src/process.rst [hunk 8]
```diff
@@ -92,7 +92,20 @@ Data types
...
-            UV_PROCESS_WINDOWS_FILE_PATH_EXACT_NAME = (1 << 7)
+            UV_PROCESS_WINDOWS_FILE_PATH_EXACT_NAME = (1 << 7),
+            /*
+             * Run the subprocess as administrator. This option is only meaningful on
+             * Windows systems. On Unix it is silently ignored.
+             * This flag is mutually exclusive with UV_PROCESS_DETACHED. UV_EINVAL will
+             * be returned if both are specified.
+             * This flag is not support stdio, cwd, env and stdio_count options in
+             * the uv_process_options_t struct, and will get UV_EINVAL if any of them
+             * are specified.
```
Comment (bnoordhuis) [inline docs/src/process.rst:103]:
```suggestion
             * This flag is mutually exclusive with UV_PROCESS_DETACHED. UV_EINVAL is
             * returned when both are specified.
             * This flag is mutually exclusive with the stdio, cwd, env and stdio_count options in
             * the uv_process_options_t struct. UV_EINVAL is returned when any of them
             * are specified.
```

File: src/win/process.c [hunk 9]
```diff
@@ -929,7 +937,21 @@ int uv_spawn(uv_loop_t* loop,
...
+       * set, we return UV_EINVAL.
+       */
+      (((options->flags & UV_PROCESS_DETACHED) ||
+        options->stdio != NULL ||
+        options->cwd != NULL ||
+        options->env != NULL ||
+        options->stdio_count > 0) &&
+       (options->flags & UV_PROCESS_WINDOWS_RUNAS_ADMIN))) {
     return UV_EINVAL;
   }
```
Comment (bnoordhuis) [inline src/win/process.c:956]: The logic here becomes hard to follow. Suggestion: break it up in several `if` statements. I can't post it as a diff but something like:
```c
if (options->file == NULL || options->args == NULL)
  return UV_EINVAL;

if (options->flags & UV_PROCESS_WINDOWS_RUNAS_ADMIN) {
  if (options->flags & UV_PROCESS_DETACHED)
    return UV_EINVAL;
  if (options->stdio != NULL || options->cwd != NULL || options->env != NULL)
    return UV_EINVAL;
  if (options->stdio_count > 0)
    return UV_EINVAL;
}
```

File: src/win/process.c [hunk 10]
```diff
@@ -1073,27 +1085,65 @@ int uv_spawn(uv_loop_t* loop,
...
+    shell_execute_info_w.cbSize = sizeof(shell_execute_info_w);
+    shell_execute_info_w.fMask = flag_mask | SEE_MASK_NOCLOSEPROCESS; /* for shell_execute_info_w.hProcess */
+    shell_execute_info_w.lpVerb = L"runas";
+    shell_execute_info_w.lpFile = application_path;
+    shell_execute_info_w.lpParameters = arguments;
+    shell_execute_info_w.nShow = startup.wShowWindow;
+    if (!ShellExecuteExW(&shell_execute_info_w)) {
+        err = GetLastError();
+        goto done;
+    }
```
Comment (bnoordhuis) [inline src/win/process.c:1098]:
```suggestion
    if (!ShellExecuteExW(&shell_execute_info_w)) {
      err = GetLastError();
      goto done;
    }
```

File: src/win/process.c [hunk 11]
```diff
@@ -1073,27 +1085,65 @@ int uv_spawn(uv_loop_t* loop,
...
+                        NULL,
+                        NULL,
+                        1,
+                        process_flags,
+                        env,
+                        cwd,
+                        &startup,
+                        &info)) {
+        err = GetLastError();
+        goto done;
```
Comment (bnoordhuis) [inline src/win/process.c:1134]:
```suggestion
      err = GetLastError();
      goto done;
```

**Review by bnoordhuis — APPROVED** [review]
LGTM with some suggestions but it'd be good if @vtjnash takes another look as well.

File: src/win/process.c (hunk 9 above)
Comment (CGQAQ) [reply src/win/process.c:956]: got it

**Comment by CGQAQ** [comment]
Hello, @vtjnash , could you please take a look?
